bip-0346.md

```diff
@@ -0,0 +1,438 @@
+  budget ("sigops budget") introduced in BIP-0342, each TransactionHash
+  decreases the validation budget by 25. If this brings the budget below zero,
+  the script fails immediately.<br>The following considerations should be made:
+  * All fields that can be of arbitrary size are cachable as TransactionHash
+    always hashes their hashed values.
+  * In "individual mode", a user can at most commit 32 inputs or outputs,
+    which we don't consider excessive for potential repeated use.
+  * In "leading mode", a caching strategy can be used where the SHA256 context
+    is stored every N in/outputs so that multiple executions of the
+    TransactionHash function can use the caches and only have to hash an
+    additional N-1 items at most.
+
+
+# Motivation
+
+This BIP specifies a basic transaction introspection primitive that is useful
+to either reduce interactivity in multi-user protocols or to enforce some basic
+constraints on transactions.
+
+Additionally, the constructions specified in this BIP can lay the groundwork for
+some potential future upgrades:
+* The TxFieldSelector construction would work well with a hypothetical opcode
+  `OP_TX` that allows for directly introspecting the transaction by putting the
+  fields selected on the stack instead of hashing them together.
+* The TransactionHash obtained by `OP_TXHASH` can be combined with a
+  hypothetical opcode `OP_CHECKSIGFROMSTACK` to effectively create an
+  incredibly flexible signature hash, which would enable constructions like
+  `SIGHASH_ANYPREVOUT`.
+
+
+## Comparing with some alternative proposals
+
+* This proposal strictly generalizes BIP-119's `OP_CHECKTEMPLATEVERIFY`, as the
+  default mode of our TxFieldSelector is semantically the same (though not
+  byte-for-byte identical) as what `OP_CTV` accomplishes, without costing any
+  additional bytes. Additionally, using `OP_CHECKTXHASHVERIFY` allows for more
+  flexibility which can help in the case for
+  * enabling adding fees to a transaction without breaking a multi-tx protocol;
+  * multi-user protocols where users are only concerned about their own inputs and outputs.
+
+* Constructions like `OP_IN_OUT_VALUE` used with `OP_EQUALVERIFY` can be
+  emulated by two `OP_TXHASH` instances by using the TxFieldSelector to select
+  a single input value first and a single output value second and enforcing
+  equality on the hashes. Neither of these alternatives can be used to enforce
+  small value differencials without the availability of 64-bit arithmetic in
+  Script.
+
+* Like mentioned above, `SIGHASH_ANYPREVOUT` can be emulated using `OP_TXHASH`
+  when combined with `OP_CHECKSIGFROMSTACK`: 
+  `<txfs> OP_TXHASH <pubkey> OP_CHECKSIGFROMSTACK` effectively emulates `SIGHASH_ANYPREVOUT`.
+
+
+# Detailed Specification
+
+A reference implementation in Rust is provided attached as part of this BIP
+together with a JSON file of test vectors generated using the reference
+implementation.
+
+
+# Backwards Compatibility
```

Comment:
This might also be a good place to mention the interactions with BIP 118, BIP 119, BIP 347, and BIP 348.